S_02/run.ipynb

```diff
@@ -124,12 +124,15 @@
     "4. Read in the Hamiltonian (as done in [S 1](../S_01/run.ipynb)) and create a Monkhorst-Pack grid as the one in Siesta input. Use this snippet to calculate the PDOS using `sisl`\n",
     "\n",
     "       mp = sisl.MonkhorstPack(H, [nx, ny, 1])\n",
-    "       mp_avg = mp.apply.average\n",
-    "       pdos = mp_avg.PDOS(E, eta=True)\n",
+    "       def wrap_multiple(eigenstate):\n",
+    "           DOS = eigenstate.DOS(E)\n"
+    "           PDOS = eigenstate.PDOS(E)\n"
+    "           return sisl.oplist([DOS, PDOS])\n",
+    "       DOS, PDOS = mp.apply.average.eigenstate(wrap=wrap_multiple, eta=True)\n",
```

Comment:
Is there a reason you amended it to also calculate the DOS?
In fact this can be much simpler:
```python
pdos = mp.apply.average(wrap=lambda es: es.PDOS(E)`
dos = pdos.sum((0, 1))
```
since DOS and PDOS are directly related.
so no need to

Reply:
Well, I guess that my thinking was about letting the students know that they may also request directly the total DOS via `DOS = eigenstate.DOS(E)`. Surely, the DOS may be obtained afterwards from the PDOS as you mentioned. However, most of the times the user might simply want the total DOS, and I guess in such a case it might be useful to know that you can simply do `eigenstate.DOS(E)`. In any case, this is a minor thing, so, @zerothi, feel free to choose what makes more sense to you! :wink:

Reply:
Besides... I was completely unaware of this option:

```python
pdos = mp.apply.average(wrap=lambda es: es.PDOS(E)`
dos = pdos.sum((0, 1))
```
What is `lambda` here? Is this a new implementation of *v0.15.0* or was already there? Such type of notation did certainly not appeared in the BZ API..

Reply:
I am a bit hesitant to show two return values if they are not used. It typically causes confusion ;)

`lambda` is a Python intrinsic (search for `lambda functions`).

Reply:
> I am a bit hesitant to show two return values if they are not used. It typically causes confusion ;)

Fair enough... then better delete it :smirk:

> lambda is a Python intrinsic (search for lambda functions).

Ok, I'll take a look as soon as I can. I guess this is not really explained anywhere in the docs, right?

Reply:
No, and it shouldn't it isn't sisl specific :)

Reply:
well, yeah, but it is a way to use the `apply` method that is not really covered in the BZ docs... anyways! not really important (at least not now..) - so, you may go on with the corrections you mention, and the merging :smile:

Reply:
Ok, I kept it and added another point.